REGRESSION (267818@main): ~6% regression on `nytimes-article` in PLT

whsieh opened the pull request:
#### 18a4511bc7e2df666e3a46b8032e7b2a823cea06
<pre>
REGRESSION (267818@main): ~6% regression on `nytimes-article` in PLT
<a href="https://bugs.webkit.org/show_bug.cgi?id=263727">https://bugs.webkit.org/show_bug.cgi?id=263727</a>
rdar://116463015

Reviewed by Tim Horton.

After the changes in 267818@main, the `nytimes-article` subtest in PLT is regressed by roughly 6% on
some iPhone device models. From running several A/B experiments, this is due to the fact that the
opportunistic task scheduler now installs a repeating `CFRunLoopTimer`; even if the runloop timer
callback is a no-op, the overhead alone seems to be enough to impact NYT article pages in
particular.

To fix this, we instead turn the runloop timer back into a one-shot timer, and only schedule it if
needed (i.e., only if opportunistic GC/sweeping is enabled, or if there are scheduled idle
callbacks).

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::requestIdleCallback):

Set a flag when calling into `requestIdleCallback`, to let the `OpportunisticTaskScheduler` know
that it needs to be scheduled for the purposes of firing opportunistic idle callbacks.

* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::rescheduleIfNeeded):

Implement the main fix here; see above for more details.

(WebCore::OpportunisticTaskScheduler::checkedPage const):

Add a helper to return a `CheckedPtr&lt;Page&gt;`; take this opportunity to deploy smart pointers (mostly,
`CheckedPtr` in this class as well).

(WebCore::OpportunisticTaskScheduler::isPageInactiveOrLoading const):

Pull this common logic out into a helper method to avoid duplicating the check.

(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):
(WebCore::OpportunisticTaskScheduler::reschedule): Deleted.

Rename this to `rescheduleIfNeeded` to reflect the fact that it may not reschedule the runloop timer
if there are no idle callbacks, and opportunistic GC/sweeping is disabled.

* Source/WebCore/page/OpportunisticTaskScheduler.h:
(WebCore::OpportunisticTaskScheduler::willQueueIdleCallback):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::renderingUpdateCompleted):

Canonical link: <a href="https://commits.webkit.org/269833@main">https://commits.webkit.org/269833@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4f9f7e152e3c7a8a1780f7c286bb79512187b9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23697 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25852 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21872 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23966 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24228 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22418 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26448 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21417 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21624 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21687 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1116 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18814 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1099 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5668 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1528 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/OpportunisticTaskScheduler.cpp:
```diff
@@ -98,14 +117,16 @@ void OpportunisticTaskScheduler::runLoopObserverFired()
+    }
 
-    if (!page->settings().opportunisticSweepingAndGarbageCollectionEnabled())
+    if (!checkedPage()->settings().opportunisticSweepingAndGarbageCollectionEnabled())
```
We may want to store the page in a local variable to avoid repeated calls to checkedPage().

whsieh committed:
REGRESSION (267818@main): ~6% regression on `nytimes-article` in PLT
https://bugs.webkit.org/show_bug.cgi?id=263727
rdar://116463015

Reviewed by Tim Horton.

After the changes in 267818@main, the `nytimes-article` subtest in PLT is regressed by roughly 6% on
some iPhone device models. From running several A/B experiments, this is due to the fact that the
opportunistic task scheduler now installs a repeating `CFRunLoopTimer`; even if the runloop timer
callback is a no-op, the overhead alone seems to be enough to impact NYT article pages in
particular.

To fix this, we instead turn the runloop timer back into a one-shot timer, and only schedule it if
needed (i.e., only if opportunistic GC/sweeping is enabled, or if there are scheduled idle
callbacks).

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::requestIdleCallback):

Set a flag when calling into `requestIdleCallback`, to let the `OpportunisticTaskScheduler` know
that it needs to be scheduled for the purposes of firing opportunistic idle callbacks.

* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::rescheduleIfNeeded):

Implement the main fix here; see above for more details.

(WebCore::OpportunisticTaskScheduler::checkedPage const):

Add a helper to return a `CheckedPtr<Page>`; take this opportunity to deploy smart pointers (mostly,
`CheckedPtr` in this class as well).

(WebCore::OpportunisticTaskScheduler::isPageInactiveOrLoading const):

Pull this common logic out into a helper method to avoid duplicating the check.

(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):
(WebCore::OpportunisticTaskScheduler::reschedule): Deleted.

Rename this to `rescheduleIfNeeded` to reflect the fact that it may not reschedule the runloop timer
if there are no idle callbacks, and opportunistic GC/sweeping is disabled.

* Source/WebCore/page/OpportunisticTaskScheduler.h:
(WebCore::OpportunisticTaskScheduler::willQueueIdleCallback):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::renderingUpdateCompleted):

Canonical link: https://commits.webkit.org/269833@main
